java/memory/memory-core/src/main/java/org/apache/arrow/memory/Accountant.java

```diff
@@ -169,9 +169,14 @@ public boolean forceAllocate(long size) {
    */
   private AllocationOutcome.Status allocate(final long size, final boolean incomingUpdatePeak,
       final boolean forceAllocation, AllocationOutcomeDetails details) {
-    final long newLocal = locallyHeldMemory.addAndGet(size);
+    final long oldLocal = locallyHeldMemory.getAndAdd(size);
+    final long newLocal = oldLocal + size;
+    // Borrowed from Math.addExact (but avoid exception here)
```

Comment:
Is it necessary to postpone exception throws?

Reply:
Exception handlers may have a performance penalty (even if not invoked), and this code is meant to be fast.